File: src/main/java/com/github/duync/jmeterviewer/JMeterFileType.java
package com.github.duync.jmeterviewer;

import com.intellij.openapi.fileTypes.LanguageFileType;
import com.intellij.openapi.util.NlsContexts.Label;
import com.intellij.openapi.vfs.VirtualFile;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.Icon;

public final class JMeterFileType extends LanguageFileType {
    public static final JMeterFileType INSTANCE = new JMeterFileType();

    private JMeterFileType() {
        super(JMeterLanguage.INSTANCE);
    }

    @Override
    public @NonNls @NotNull String getName() {
        return "JMeter Test Plan";
    }

    @Override
    public @Label @NotNull String getDescription() {
        return "Apache JMeter test plan";
    }

    @Override
    public @NotNull String getDefaultExtension() {
        return "jmx";
    }

    @Override
    public @Nullable Icon getIcon() {
        return null;
    }

    @Override
    public boolean isReadOnly() {
        return false;
    }

    @Override
    public @Nullable String getCharset(@NotNull VirtualFile file, byte @NotNull [] content) {
        return "UTF-8";
    }
}

File: src/main/kotlin/com/github/kohebth/jmeterviewer/ide/JMeterFileType.kt
package com.github.kohebth.jmeterviewer.ide

import com.intellij.lang.xml.XMLLanguage
import com.intellij.openapi.fileTypes.LanguageFileType
import com.intellij.openapi.vfs.VirtualFile
import java.nio.charset.StandardCharsets
import javax.swing.Icon
import javax.swing.ImageIcon

object JMeterFileType : LanguageFileType(XMLLanguage.INSTANCE) {
    private val jmeterIcon = ImageIcon(
        checkNotNull(JMeterFileType::class.java.getResource("/icons/jmeter-feather.gif")) {
            "Missing bundled JMeter feather icon"
        },
    )

    override fun getName(): String = "JMeter Test Plan"

    override fun getDescription(): String = "Apache JMeter test plan"

    override fun getDefaultExtension(): String = "jmx"

    override fun getIcon(): Icon = jmeterIcon

    override fun getCharset(file: VirtualFile, content: ByteArray): String =
        StandardCharsets.UTF_8.name()
}
